src/adsbinarynode.jl

```diff
@@ -0,0 +1,402 @@
+    mask = SVector{N, Bool}(s.lo[i] <= x[i] <= s.hi[i] for i in 1:N)
+    return all(mask)
+end
+
+"""
+    ADSBinaryNode{N, M, T <: Real}
+
+Automatic Domain Splitting binary tree with `N` jet transport variables
+and `M` degrees of freedom.
+
+# Fields
+
+- `s::ADSDomain{N, T}`: domain of jet transport variables.
+- `t::T`: time.
+- `x::SVector{M, TaylorN{T}}`: state vector at time `t`.
+- `p::SVector{M, Taylor1{TaylorN{T}}}`: polynomial dependence of state vector wrt time.
+- `depth::Int`: node depth.
+- `parent::Union{Nothing, ADSBinaryNode{N, M, T}}`: parent node.
+- `left::Union{Nothing, ADSBinaryNode{N, M, T}}`: left child.
+- `right::Union{Nothing, ADSBinaryNode{N, M, T}}`: right child.
+
+!!! reference
+    This implementation is based upon the **binarytree.jl** example at
+    https://github.com/JuliaCollections/AbstractTrees.jl/blob/master/test/examples/binarytree.jl
+"""
+mutable struct ADSBinaryNode{N, M, T <: Real}
+    s::ADSDomain{N, T}
+    t::T
+    x::SVector{M, TaylorN{T}}
+    p::SVector{M, Taylor1{TaylorN{T}}}
+    depth::Int
+    parent::Union{Nothing, ADSBinaryNode{N, M, T}}
+    left::Union{Nothing, ADSBinaryNode{N, M, T}}
+    right::Union{Nothing, ADSBinaryNode{N, M, T}}
+    # Inner constructor (root node by default)
+    function ADSBinaryNode{N, M, T}(
+        s::ADSDomain{N, T}, t::T, x::SVector{M, TaylorN{T}},
+        p::SVector{M, Taylor1{TaylorN{T}}}, depth::Int = 0,
+        parent = nothing, l = nothing, r = nothing
+        ) where {N, M, T <: Real}
+        new{N, M, T}(s, t, x, p, depth, parent, l, r)
+    end
+end
+
+# Outer constructor (root node)
+function ADSBinaryNode(
+    s::ADSDomain{N, T}, t::T, x::SVector{M, TaylorN{T}},
+    p::SVector{M, Taylor1{TaylorN{T}}}
+    ) where {N, M, T <: Real}
+    return ADSBinaryNode{N, M, T}(s, t, x, p)
+end
+
+# Print method for ADSBinaryNode
+# Example:
+# x: [1.0, 0.0, 0.0, 1.224744871391589]
+# s: [-1.0, 1.0]×[-1.0, 1.0]
+# t: 0.0
+function show(io::IO, n::ADSBinaryNode{N, M, T}) where {N, M, T <: Real}
+    print(io, "x: ", constant_term.(n.x), "\ns: ", n.s, "\nt: ", n.t)
+end
+
+# Left (right) child constructors
+function leftchild!(
+    parent::ADSBinaryNode{N, M, T}, s::ADSDomain{N, T}, t::T,
+    x::SVector{M, TaylorN{T}}, p::SVector{M, Taylor1{TaylorN{T}}}
+    ) where {N, M, T <: Real}
+    isnothing(parent.left) || error("Left child is already assigned")
+    node = ADSBinaryNode{N, M, T}(s, t, x, p, parent.depth + 1, parent)
+    parent.left = node
+end
+
+function rightchild!(
+    parent::ADSBinaryNode{N, M, T}, s::ADSDomain{N, T}, t::T,
+    x::SVector{M, TaylorN{T}}, p::SVector{M, Taylor1{TaylorN{T}}}
+    ) where {N, M, T <: Real}
+    isnothing(parent.right) || error("Right child is already assigned")
+    node = ADSBinaryNode{N, M, T}(s, t, x, p, parent.depth + 1, parent)
+    parent.right = node
+end
+
+# AbstractTrees interface
+function children(node::ADSBinaryNode{N, M, T}) where {N, M, T <: Real}
```

Comment:
For all the method overloads from `Base` and `AbstractTrees` I'd suggest to switch to explicitly prefixing the import so that it's clear which methods we are overloading from where. E.g.,

```julia
import Base, AbstractTrees

function AbstractTrees.children(...)
    # (overload for ADSBinary Node)
end

function AbstractTrees.printnode(...)
    # (overload for ADSBinaryNode)
end

function Base.IteratorEltype(...)
    # overload for ADSBinaryNode
end

# etc...
```